support quadword only flash in storage

TychoVrahe opened the pull request:
Implements support for quadword-only flash in storage. 

Currently done via conditional compilation inside the C storage and most of the tests are run against both variants in loop. Python implementation is selected on object creation time. The differences are relatively large, so not sure that this organization is the best, maybe it would better/easier to read if we split the two implementations entirely? 

Upgrade tests are not implemented, as upgrading from other storage versions does not make sense.




<!--
If you are a core dev:

Don't forget to set up the fields:
- assign yourself
- set priority to same as original issue
- add PR to sprint
- if Draft -> "in progress"
- if final PR -> "needs review"

If you're an external contributor, you can ignore this text.
-->

Hannsek commented:
@andrewkozlik do you have any estimations when you can take a look at this?

andrewkozlik reviewed (commented):
Partial review, as discussed IRL.

TychoVrahe commented on storage/tests/python/tests/test_norcow_qw.py:
```diff
@@ -0,0 +1,305 @@
+from . import common
+
+
+def test_norcow_set_qw():
```
added `test_norcow_update` that sets and updates items with len 11, 12, and 13: https://github.com/trezor/trezor-firmware/commit/5b234749c62d89486630160bf4835ddef5e068cd

TychoVrahe commented on storage/norcow.c:
```diff
@@ -502,18 +739,33 @@ secbool norcow_delete(uint16_t key) {
+#ifdef FLASH_BYTE_ACCESS
+    ensure(flash_area_write_word(area, offset, 0x00000000), NULL);
+#else
+    if (len >= 12) {
```
i confirm this was an error. fixed by you in [suggestions3](https://github.com/trezor/trezor-firmware/commit/e7dd0661735f1018f63212f74ad79a3b1135b189) . 

A added a test that would have caught this in https://github.com/trezor/trezor-firmware/commit/95c046a7b9f1a48716bf8b839507599f44428133

TychoVrahe commented on storage/norcow.c:
```diff
@@ -39,10 +37,26 @@
+#define NORCOW_WORD_SIZE (4 * sizeof(uint32_t))
+#define NORCOW_PREFIX_LEN (2 * NORCOW_WORD_SIZE)
+#define NORCOW_KEY_LEN 16
+#define NORCOW_LEN_LEN 16
```
outdated since [suggestions2](https://github.com/trezor/trezor-firmware/commit/b29f78b30bf9e503fa01892c92a01cff1191b130)

TychoVrahe commented on storage/storage.c:
```diff
@@ -404,19 +414,22 @@ static secbool set_wipe_code(const uint8_t *wipe_code, size_t wipe_code_len) {
+         WIPE_CODE_SALT_SIZE + WIPE_CODE_TAG_SIZE);
+
+  // Write wipe code into the preallocated entry.
+  if (sectrue != norcow_update_bytes(WIPE_CODE_DATA_KEY, 0, (uint8_t *)data,
```
ack [suggestions](https://github.com/trezor/trezor-firmware/commit/ae64cab7e5144bbfcebb8e109cb44e8d1f7c80a6)

TychoVrahe commented on storage/storage.c:
```diff
@@ -126,6 +126,12 @@ const uint32_t V0_PIN_EMPTY = 1;
+#ifdef FLASH_BYTE_ACCESS
+#define CHACHA20_IV_PADDING 0
+#else
+#define CHACHA20_IV_PADDING 4
```
Due to encrypted data are organized, the buffering approach does not work (unless the buffer is same size as the as the whole item size, but then using `update_bytes` is useless + we don't have any sort of dynamic allocation on this level so its problematic).

The reason is that the `TAG` is stored after the `IV` at the beginning, but is computed and written after the encrypted data. Therefore, the IV and (beginning of) the message would have to be stored until the tag is computed. In more general situation, we cannot tell anything about the order of data fed into update_bytes, therefore, we would have to buffer the whole thing.

One way out is to change the update_bytes function so that it does not accept offset as a parameter, but collects the data sequentially: but that require storing the tag after the encrypted data, which is incompatible with currently stored data. Which leads to upgrading the storage, or keeping two versions of encrypted data. 

Also this would limit the functionality of the `update_bytes` function. 

This is to some extent related to the comment above related to the data structure of full-size items above. If we add the 14 bytes of data to the same quadword as key, we need to buffer the key until the data is known.

andrewkozlik replied:
Good catch. I forgot that we put the tag at the beginning. That wasn't a very good choice.
> One way out is to change the update_bytes function so that it does not accept offset as a parameter, but collects the data sequentially

That's what I assumed we'd do, but as you correctly note, it's incompatible with currently stored data.

TychoVrahe commented on storage/norcow.c:
```diff
@@ -100,6 +117,85 @@ static secbool norcow_write(uint8_t sector, uint32_t offset, uint32_t prefix,
   return sectrue;
 }
+#else
+static secbool norcow_write(uint8_t sector, uint32_t offset, uint16_t key,
+                            const uint8_t *data, uint16_t len) {
```
The else branch is somewhat problematic because of the udpate_bytes mechanism - can't write the key until enough data has been pushed through update_bytes, but update_bytes will not find the item in storage until the key is written. 

We could do something like this in the else branch:

- First quadword:
     - 2 bytes len
     - 2 bytes key
- Second quadword:
     - 1 byte deleted flag 
     - 15 bytes data

Is it ok to have the key remain in the first QW after deletion?

TychoVrahe replied:
There are additional issues with the else branch (even with the updated version):
- once you start streaming the bytes, you have to finish before streaming some other key, otherwise it fails (this is not broken anywhere at this moment, but its a new limitation that one needs to be aware of)
- the current implementation of storage's `secequal32` expects the data aligned to 32bits. This is also breaking boundaries between storage and norcow as its making an assumption about how the data are organized. We can either remove the asserts (our current MCU's can handle unaligned access with some performance penalty, which i deem acceptable at this situation), or we can perhaps use `secequal` instead. 

With that said, here is implementation of this optimization and also the update_bytes buffering: https://github.com/trezor/trezor-firmware/commit/51666699faa5e6735147fa7927cddd37a8d81877

Note that test_norcow_qw are currently failing as i didn't to waste more time until we decide that this is a way to go. Other tests are ok. Also the data structure is incompatible with older storages as mentioned below: we could either have two variants or do a storage upgrade.

TychoVrahe replied:
In subsequent commits https://github.com/trezor/trezor-firmware/pull/3228/commits/d278e20f8f1181cea9a8f289b1986f2f60c5968c and https://github.com/trezor/trezor-firmware/pull/3228/commits/d3de9942eb080b1ab29f25f22906e53185e3c16c
I have merged the pin logs behavior in storage.c: now it doesn't use `update_word` but always calls norcow_set, as this function is able to determine what is updatable and what isn't.
And counter implementation is moved to norcow.c, because how it is implemented is dependent on underlying hw. 

With these changes, storage.c is now independent on how norcow.c is implemented

andrewkozlik commented:
I did some refactoring to split the implementations of the NORCOW functions that differ based on `FLASH_BIT_ACCESS`. Having all the code mixed together made it illegible for me. I'd just get lost in the ifdefs. There is still some work to do:
- It's completely untested. I might have broken something in the refactoring.
- `FLASH_BLOCK_WORDS` needs to be defined in the build scripts as `1` for T1, TT and TS3.
- I still need to review the code thoroughly, but right now the concept is beginning to look good to me. 
- We can get rid of the buffering mechanism in `norcow_update_bytes()`, which is non-trivial, if we just specify and assert that the `len` parameter must be a multiple of `FLASH_BLOCK_SIZE`, except when it's the final update. For that we just need to move the `NORCOW_VALID_FLAG` to the end, i.e. the format would be: value length 2 bytes, key 2 bytes, padding, value data, `NORCOW_VALID_FLAG`, padding.

TychoVrahe commented:
Thanks, i have added a commit with additional fixes. Tests are now succeeding (apart from the norcow_qw test, which i will fix when the data format is finalized). Not tested on HW yet.

I have added some additional tests for things that were previously broken.

The logic when the item update is possible is more complete, it allows update for - equal data, updating the whole quadword/block to all 0x00, updating to any value if previously the quadword/block was all 0xFF. 

I have changed the padding to 0xFF in some cases, as this allows updating in some additional cases, but more importantly there is an edge case: when setting encrypted item if norcow runs out of space, compacts, it writes the incomplete item. If its padded with zeros, when finalizing the item via update_bytes this fails because the last quadword is not virgin anymore.

Note that the flash_common.c gets refactored somewhat in successive PR implementing the U5 support (https://github.com/trezor/trezor-firmware/pull/3370) so some of the stuff done here will be different after that.

Also for the firmware updates using flash_block: we intend to use flash burst (8*quadword) programming (instead of block/quadword) on U5 for fw updates because its significantly faster - so the changes done in https://github.com/trezor/trezor-firmware/pull/3228/commits/a512312aac50f081a3d55f5a0ca39433f0aa177c will also not live long.

I think that getting rid of the buffering like suggested will shift some of the complexity to storage.c, since the `set_encrypted` will have to worry about the blocks being full, i.e. `CHACHA IV` padding will have to be readded, or we will have to rearrange the block to contain the IV and firsts few bytes of data. And if there is ever need to use update_bytes elsewhere, you will also need to worry about this there. So i would prefer the complexity being hidden like it is now.

andrewkozlik commented on storage/norcow_blockwise.h:
```diff
@@ -72,11 +72,10 @@ static secbool write_item(uint8_t sector, uint32_t offset, uint16_t key,
       uint16_t len_to_write =
           len > NORCOW_DATA_OPT_SIZE ? NORCOW_DATA_OPT_SIZE : len;
-      memset(block, 0, sizeof(block));
-      block[0] = NORCOW_VALID_FLAG;
```
We should write the valid flag always. If the definition of `NORCOW_VALID_FLAG` changes or we start `memset`ing to something else, then the code will break.

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/commit/de177fd1aa86751e05d2fae01e647f4937054315

andrewkozlik commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,308 @@
+  ensure(flash_lock_write(), NULL);
+}
+
+static secbool flash_area_write_bytes(const flash_area_t *area, uint32_t offset,
+                                      uint16_t dest_len, const void *val,
+                                      uint16_t len) {
+  if (val == NULL || dest_len != len) {
+    return secfalse;
+  }
+
+  uint8_t *data = (uint8_t *)val;
+
+  while (len > 0) {
+    flash_block_t block = {0};
+
+    uint16_t offset_in_block = offset % FLASH_BLOCK_SIZE;
+
+    memcpy(block,
+           flash_area_get_address(area, offset - offset_in_block,
+                                  FLASH_BLOCK_SIZE),
+           FLASH_BLOCK_SIZE);
+
+    int32_t len_to_write = len > (FLASH_BLOCK_SIZE - offset_in_block)
+                               ? (FLASH_BLOCK_SIZE - offset_in_block)
+                               : len;
+
+    memcpy(&((uint8_t *)block)[offset_in_block], data, len_to_write);
+
+    if (sectrue !=
+        flash_area_write_block(area, offset - offset_in_block, block)) {
+      return secfalse;
+    }
+
+    offset += len_to_write;
+    data += len_to_write;
+    len -= len_to_write;
+  }
+
+  return sectrue;
+}
```
I think it doesn't make sense to implement this function for the blockwise flash access, because the only use-case that I am aware of is updating the PIN counter in storage, which wouldn't work with this anyway. In fact, we should get rid of the hardened PIN counter implementation at least for the U5 and maybe even for the older models.

TychoVrahe replied:
One example when this actually applies is when changing pin: the `PIN_NOT_SET_KEY` item gets overwritten with same value, so if this is not implemented the value gets deleted and re-added in norcow. Maybe its not worth implementing anyways, also this specific case can be probably handled on `storage.c` level. On the other hand, for similar future situations this might be good to keep. I will let you decide, i think its fine either way and if go without this, readding it in future, if the need arises, should not be a problem i believe.

andrewkozlik replied:
> the PIN_NOT_SET_KEY item gets overwritten with same value,
> On the other hand, for similar future situations this might be good to keep.

I think that those situations can be covered by something like this:
```c
return memcmp(val, flash_area_get_address(area, offset, dest_len), len) == 0 ? sectrue : secfalse;
```
Any additional complexity seems unwarranted.

TychoVrahe replied:
using simplified version in https://github.com/trezor/trezor-firmware/commit/f625f87fb56badc70fb6a47ff6ab4574aad0aed6

andrewkozlik commented on storage/norcow_blockwise.h:
```diff
@@ -236,9 +262,8 @@ secbool norcow_update_bytes(const uint16_t key, const uint8_t *data,
-    memset(norcow_write_buffer, 0, sizeof(norcow_write_buffer));
+    memset(norcow_write_buffer, 0xFF, sizeof(norcow_write_buffer));
     norcow_write_buffer_key = key;
-    norcow_write_buffer[0] = NORCOW_VALID_FLAG;
```
Here too, we should always write the `NORCOW_VALID_FLAG` and not rely on the identity with the memset. The compiler will optimize the code anyway if the two are identical.

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/commit/de177fd1aa86751e05d2fae01e647f4937054315

andrewkozlik commented on storage/flash_common.c:
```diff
@@ -115,13 +117,57 @@ secbool flash_area_write_word(const flash_area_t *area, uint32_t offset,
+  secbool new_all_00 = sectrue;
+  secbool all_equal = sectrue;
+
   for (int i = 0; i < 4; i++) {
```
4 should be `FLASH_BLOCK_WORDS`

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/commit/6e1e42d6e01cc267862ef3571581a548030d7dab

andrewkozlik commented on storage/flash_common.c:
```diff
@@ -115,13 +117,57 @@ secbool flash_area_write_word(const flash_area_t *area, uint32_t offset,
+    }
+  }
+
+  if (!(old_all_ff == sectrue || new_all_00 == sectrue ||
+        all_equal == sectrue)) {
+    return secfalse;
+  }
```
Why do we need this? Shouldn't the underlying implementation of `flash_write_word()` be doing this internally? Also, if `FLASH_BIT_ACCESS` is true, then this is too restrictive. (I think `FLASH_BIT_ACCESS` and `FLASH_BLOCK_WORDS` should be treated as independent.)

TychoVrahe replied:
this will be part of U5 flash driver implementation (specifically its `flash_write_quadword` function) when we have one (its part of https://github.com/trezor/trezor-firmware/pull/3370 ), but i also need something here to make tests work. Probably part of `tests/c/flash.c` is better place though? (compiled conditionally depending on the `FLASH_BIT_ACCESS` and `FLASH_BLOCK_WORDS` )

andrewkozlik replied:
In that case it definitely belongs into tests/c/flash.c.

TychoVrahe replied:
moved https://github.com/trezor/trezor-firmware/commit/31ca162522180b08acda83e0f00f8e615f1e8999

andrewkozlik commented on storage/tests/c/flash.h:
```diff
@@ -24,7 +24,9 @@
 #include "secbool.h"
 
-#define FLASH_BYTE_ACCESS 1
+#ifndef FLASH_QUADWORD
```
I think `FLASH_BIT_ACCESS` and `FLASH_QUADWORD` (which BTW is equivalent to `FLASH_BLOCK_WORDS==4`) should be treated as independent and should both be specified in the build flags rather than being inferred from one another.

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/commit/31ca162522180b08acda83e0f00f8e615f1e8999

andrewkozlik commented on storage/tests/python/src/consts.py:
```diff
@@ -121,6 +120,7 @@
 
 # The length of the ChaCha20 IV (aka nonce) in bytes as per RFC 7539.
 CHACHA_IV_SIZE = 12
+CHACHA_IV_PADDING = 4
```
Seems unused.

TychoVrahe replied:
removed https://github.com/trezor/trezor-firmware/commit/3fd5e9b4e90f99ee3ba85e1749db85cf0326730c

andrewkozlik commented on storage/tests/tests/test_compact.py:
```diff
@@ -6,27 +6,31 @@
-        s.set(0x07FE, b"worxxxxxxxxxxxxxxxxxx")
-        s.set(0x09EE, b"worxxxxxxxxxxxxxxxxxx")
-    assert common.memory_equals(sc, sp)
+    for byte_access in (
+        True,
+        False,
```
Is there any reason not to use `@pytest.mark.parametrize` like we do in device tests?

TychoVrahe replied:
refactored https://github.com/trezor/trezor-firmware/commit/7ad1c6ae1e81f4975045b862c627881ec1d70de3

andrewkozlik commented on storage/tests/tests/test_random_qw.py:
Let's not duplicate an entire file just because of a difference in one single line. There must be a better way of doing this.

TychoVrahe replied:
using subclasses https://github.com/trezor/trezor-firmware/commit/beb63e87fdced92fd6921642076c0962b8b7d554

whether this is actually the best way to do this, i have no idea

andrewkozlik commented on storage/tests/python/src/norcow.py:
```diff
@@ -4,29 +4,55 @@
+    return data + padding * align_int(len(data), align)
 
 
 class Norcow:
```
I think this should be implemented using two derived classes `NorcowBitwise` and `NorcowBytewise` derived from an abstract `Norcow` base class. In the tests the classes can be instantiated using a factory function for example.

TychoVrahe replied:
refactored https://github.com/trezor/trezor-firmware/commit/7ad1c6ae1e81f4975045b862c627881ec1d70de3

andrewkozlik commented on storage/tests/python/src/norcow.py:
```diff
@@ -103,64 +174,162 @@ def _append(self, key: int, value: bytes):
+                    if key == consts.NORCOW_KEY_FREE:
+                        raise ValueError("Norcow: no data on this offset")
+
+        return key, length, value
```
What's the difference between `length` and `len(value)`?

TychoVrahe replied:
In `is_updatable` in `blockwise` implementation i use the `padding` param to specify that i want to read the whole item including padding, as i need to know the whole thing to be able to determine updatability. Then l`en(value)` is the lenght of the whole item and `length` is the length of data only

andrewkozlik replied:
In that case call it `padded_value`, because this is not at all obvious. However, as discussed elsewhere, I'd rather not deal with updatability in the blockwise implementation, because it's not worth the trouble.

TychoVrahe replied:
removed with the updatability simplification https://github.com/trezor/trezor-firmware/commit/f625f87fb56badc70fb6a47ff6ab4574aad0aed6

andrewkozlik commented on storage/tests/python/src/norcow.py:
```diff
@@ -4,29 +4,55 @@
-def align4_int(i: int):
-    return (4 - i) % 4
+def align_int(i: int, align: int):
+    return (align - i) % align
```
This is weird. I'd just do `(-i) % align`.

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/commit/7b195900fe382ee4c713e131499d3eaadcb10235

TychoVrahe commented:
fix the builds with the last few commits (apart from bootloader emulator build which will be fixed on rebase, i think).

had to redefine `flash_block_t `for `FLASH_BLOCK_WORDS=1` as array too, to enable usage of this construct `&src[FLASH_BLOCK_WORDS * i]`

@andrewkozlik let me know when rebase is ok for you, there are some conflicts to solve

andrewkozlik commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,310 @@
+      uint16_t len_to_write =
+          len > NORCOW_DATA_OPT_SIZE ? NORCOW_DATA_OPT_SIZE : len;
+      memset(block, 0xFF, sizeof(block));
+      block[0] = NORCOW_VALID_FLAG | 0xFFFFFF00;
```
```suggestion
      ((uint8_t *)block)[0] = NORCOW_VALID_FLAG;
```

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/pull/3228/commits/c9366a1f4b5b81e2166402441b93187bb2593610

andrewkozlik commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,310 @@
+
+  if (norcow_write_buffer_key == -1) {
+    memset(norcow_write_buffer, 0xFF, sizeof(norcow_write_buffer));
+    norcow_write_buffer[0] = NORCOW_VALID_FLAG | 0xFFFFFF00;
```
```suggestion
    ((uint8_t *)norcow_write_buffer)[0] = NORCOW_VALID_FLAG;
```

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/pull/3228/commits/c9366a1f4b5b81e2166402441b93187bb2593610

andrewkozlik commented:
I am a little confused by the terminology. We use `sector` to refer to the index of the storage area in `STORAGE_AREAS` and then inside the storage area there are subareas which are again composed of "sectors". Since we currently use storage areas with only one subarea that has only one sector, it sort of coincides. So for example `erase_sector(uint8_t sector, ...)` should be something like `erase_storage_area(uint8_t storage_area_index, ...)`.

We work with `NORCOW_SECTOR_SIZE`, but what does it mean now that we have the areas? I think it means the area size or subarea size. The code is actually not able to handle multiple subareas, because it relies on contiguous memory. So we should have a static assert or a comment in the appropriate places stating that `num_subareas` must be equal to 1 for `STORAGE_AREAS`. It also seems to me that the code will incorrectly handle cases when we have multiple sectors in a subarea. I think it will be able to write the items, but it won't be able to read the item if it is written over multiple sectors in the subarea, because `flash_area_get_address()` doesn't take this situation into account.

Should be handled in a separate issue though.

TychoVrahe commented:
yes i agree that it would be beneficial to update `sector` terminology in storage and add the assert, lets create an issue for that.

As for the `flash_area_get_address ` not handling sector boundaries correctly, this is fixed in the subsequent [PR](https://github.com/trezor/trezor-firmware/pull/3370), see 
https://github.com/trezor/trezor-firmware/blob/36b80e8a10cf027b2050deb66ea524ce2f4b25ad/storage/flash_area.c#L62

andrewkozlik commented on storage/tests/c/flash.h:
```diff
@@ -24,7 +24,19 @@
 #include "secbool.h"
 
-#define FLASH_BYTE_ACCESS 1
+#define FLASH_BLOCK_SIZE (sizeof(uint32_t) * FLASH_BLOCK_WORDS)
+
+typedef uint32_t flash_block_t[FLASH_BLOCK_WORDS];
+
+#if FLASH_BLOCK_WORDS == 1
+#define FLASH_ALIGN(X) (((X) + 3) & ~3)
+#define FLASH_IS_ALIGNED(X) (((X)&3) == 0)
+#elif FLASH_BLOCK_WORDS == 4
+#define FLASH_ALIGN(X) (((X) + 0xF) & ~0xF)
+#define FLASH_IS_ALIGNED(X) (((X)&0xF) == 0)
+#else
+#error Unsupported number of FLASH_BLOCK_WORDS.
+#endif
```
Darn, I though I put these in flash_common.h, so that we don't have to redefine them elsewhere (core/embed/trezorhal/flash.h, legacy/flash.h). Also some of the declarations below could probably also be moved to flash_common.h, (some going under #ifdef FLASH_BIT_ACCESS).

andrewkozlik replied:
Actually the functions that are declared below should go into some internal header file, because they are not meant to be used directly anymore now that we have the `flash_area_*` functions. They are only meant for the implementation of the `flash_area_*` functions. But this particular refactoring job can be handled in a separate issue.

TychoVrahe replied:
reverted in https://github.com/trezor/trezor-firmware/commit/87ca3a5654dd208defdaf5a63216d8ea748194e1

moved the flags to build scripts from header file to make it compile.

note that `flash_common.h` and related stuff get refactored in https://github.com/trezor/trezor-firmware/pull/3370 , specifically in commit https://github.com/trezor/trezor-firmware/pull/3370/commits/757fed5ef4fa3c3c906a607b9096e0427f32ed08 , which includes creating the internal header file called `flash_ll.h`. `flash_common.h` is changed into `flash_area.h` and these new definitions will probably be moved to `flash_ll`, so the placement of these macros here is temporary anyway.

andrewkozlik commented on legacy/flash.h:
```diff
@@ -24,9 +24,18 @@
+
 #define FLASH_SECTOR_COUNT 24
 
+#define FLASH_BLOCK_SIZE (sizeof(uint32_t) * FLASH_BLOCK_WORDS)
+
+typedef uint32_t flash_block_t[FLASH_BLOCK_WORDS];
+
+#define FLASH_ALIGN(X) (((X) + 3) & ~3)
+#define FLASH_IS_ALIGNED(X) (((X)&3) == 0)
+
```
Remove once the definitions are moved to flash_common.h.

TychoVrahe replied:
reverted in https://github.com/trezor/trezor-firmware/commit/87ca3a5654dd208defdaf5a63216d8ea748194e1

andrewkozlik reviewed (changes requested):
Norcow reviewed. Let's think about where the "valid flag" should be. Still need to look at storage.c.

TychoVrahe commented on core/embed/rust/build.rs:
```diff
@@ -133,6 +140,8 @@ fn prepare_bindings() -> bindgen::Builder {
             "-I../../vendor/micropython/ports/unix",
             "-DTREZOR_EMULATOR",
+            "-DFLASH_BIT_ACCESS=1",
+            "-DFLASH_BLOCK_WORDS=1",
```
rust (specifically bindgen) needs to know these constants too. We currently don't have universal solution to pass gcc command line macros to rust, so its done manually. Developing some system that would pass all the macros to rust/bindgen automatically would be nice, but its out of scope here

onvej-sl replied:
@andrewkozlik Regarding this comment, is there anything that still needs to be done?

andrewkozlik replied:
I don't want to hold up the review with a build script cleanup issue, and I am not looking for a universal solution to pass gcc command line macros to rust. What I am not happy with are the facts that:
1. The `FLASH_*` constants are defined multiple times for no apparent reason.
2. The `FLASH_*` constants are handled differently in the emulator build than in the firmware build.

Regarding point 1: For example in site_scons/boards/stm32f4_common.py we could take the `defines` list (or probably just the three items that are added to the `defines` list) and append them to `RUST_INCLUDES` using `";".join("-D{}={}".format(k,v) for k,v in defines)` instead of manually rewriting all the defined values. Another option, although not as extendable, is to define the two values as Python variables `FLASH_BLOCK_WORDS` and `FLASH_BIT_ACCESS` and fill them in like so:
```py
   ...
    FLASH_BLOCK_WORDS = 1
    FLASH_BIT_ACCESS = 1
    defines += [
        ...
        ("FLASH_BLOCK_WORDS", f"{FLASH_BLOCK_WORDS}"),
        ("FLASH_BIT_ACCESS", f"{FLASH_BIT_ACCESS}"),
    ]
    ...
    env.get("ENV")["RUST_INCLUDES"] = (
        ...
        f"-DFLASH_BLOCK_WORDS={FLASH_BLOCK_WORDS};"
        f"-DFLASH_BIT_ACCESS={FLASH_BIT_ACCESS}"
    )
```

Regarding point 2: In the firmware build the constants are passed to the rust build via the environment variable `RUST_INCLUDES` (not a great name, but that's not important). Why can't we do it the same way in the emulator build? Why should they have to be redefined "manually" in a completely different file deep in the build system instead of being pulled from the environment variable in `embed/rust/build.rs`?

If this is more complicated than I am seeing, then let's create a separate issue for the cleanup of these variables, but it seems pretty simple to take care of now. The problem will only continue growing once we add an emulator that uses blockwise storage.

TychoVrahe replied:
While i completely agree with both points, i am against introducing half-measures here to address them only to redo it later properly. 

Neither of the issues originates here, e.g. `STM32_HAL_H` macro is defined exactly the same duplicitous way like new macros, and emulator macros being defined in `build.rs`

Ad 1)
yes this would solve this for these three constants, but to solve this properly you need to address this across all the defines across all boards, and for include paths too and those have different path, so that needs addressing too. 

Ad 2)
yes this would be desirable, but currently there is no `RUST_INCLUDES` in emulator build script, nor the `board` definitions where this originate in hw builds. We should introduce both, but its clearly out of scope here.

The separate issue has already been created (https://github.com/trezor/trezor-firmware/issues/3526) , i would say lets move on.

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+static uint16_t norcow_write_buffer_filled = 0;
+static uint16_t norcow_write_buffer_filled_data = 0;
+static int32_t norcow_write_buffer_key = -1;
+
```
added a description https://github.com/trezor/trezor-firmware/commit/3687b390832d0c8670c7c67b2c9f5619a04b18f8

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+  } else {
+    uint16_t len_adjusted = FLASH_ALIGN(len);
+
+    if (offset + NORCOW_MAX_PREFIX_LEN + len_adjusted > NORCOW_SECTOR_SIZE) {
```
fixed https://github.com/trezor/trezor-firmware/commit/f5cd4eed4b0af3472a9759a913f48f6c87c45b6e

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+    return secfalse;
+  }
+
+  flash_block_t block = {len | ((uint32_t)key << 16)};
```
Previous iteration used two quadwords/blocks for storing `len` and `key` separately, given that `len` cannot be deleted it made sense to place it first. 

Thats no longer the case so unification makes sense : https://github.com/trezor/trezor-firmware/commit/f5cd4eed4b0af3472a9759a913f48f6c87c45b6e

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+    }
+    *pos += NORCOW_KEY_LEN;
+  } else {
+    *pos += (NORCOW_KEY_LEN + NORCOW_SMALL_ITEM_SIZE);
```
fixed https://github.com/trezor/trezor-firmware/commit/f5cd4eed4b0af3472a9759a913f48f6c87c45b6e

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+    }
+
+    *pos += NORCOW_VALID_FLAG_LEN;
+    if (*((const uint8_t *)flg) == 0) {
```
fixed https://github.com/trezor/trezor-firmware/commit/f5cd4eed4b0af3472a9759a913f48f6c87c45b6e

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+}
+
+/*
+ * Update the value of the given key starting at the given offset.
```
updated https://github.com/trezor/trezor-firmware/commit/f5cd4eed4b0af3472a9759a913f48f6c87c45b6e

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+    return secfalse;
+  }
+
+  if (norcow_write_buffer_flashed + len > allocated_len) {
```
moved the valid flag to the end: https://github.com/trezor/trezor-firmware/commit/f823cb34ac3324f551de18ffad54474f0c862c0d

it presents some challenges, like a need to check whether the last value block is full or not to determine whether separate block with the flag needs to be written. The `delete_head` function now does not make much sense either, as there is no head for blockwise impl. 

Overall i would slightly favor placing the flag at the end.

andrewkozlik replied:
I agree, placing it at the end is better.

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+
+  if (norcow_write_buffer_key != key && norcow_write_buffer_key != -1) {
+    // some other update bytes is in process, abort
+    return secfalse;
```
fixed https://github.com/trezor/trezor-firmware/commit/f5cd4eed4b0af3472a9759a913f48f6c87c45b6e

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+
+    if (norcow_write_buffer_filled == FLASH_BLOCK_SIZE ||
+        (norcow_write_buffer_filled_data + norcow_write_buffer_flashed) ==
+            allocated_len + NORCOW_VALID_FLAG_LEN) {
```
https://github.com/trezor/trezor-firmware/commit/f823cb34ac3324f551de18ffad54474f0c862c0d

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,284 @@
+      ensure(flash_area_write_block(area, flash_offset, norcow_write_buffer),
+             NULL);
+      ensure(flash_area_write_block(area, flash_offset, norcow_write_buffer),
+             NULL);
```
removed https://github.com/trezor/trezor-firmware/commit/940757cb4cc218134d63af336270f0dc8adb713b

TychoVrahe commented:
modified also emulator flash driver to support different flash block sizes https://github.com/trezor/trezor-firmware/commit/2d615ea4588b17129fca271db974ef5fc36ff607

andrewkozlik commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,333 @@
+           data_to_copy);
+    data += data_to_copy;
+    norcow_write_buffer_filled += data_to_copy;
+    tmp_len -= data_to_copy;
+
+    bool all_data_received = (norcow_write_buffer_filled +
+                              norcow_write_buffer_flashed) == allocated_len;
+    bool block_full = norcow_write_buffer_filled == FLASH_BLOCK_SIZE;
+
+    if (block_full || all_data_received) {
+      if (!block_full) {
+        // all data has been received, add valid flag to last block
+        ((uint8_t *)norcow_write_buffer)[norcow_write_buffer_filled] =
+            NORCOW_VALID_FLAG;
+      }
+
+      ensure(flash_area_write_block(area, flash_offset, norcow_write_buffer),
+             NULL);
+      flash_offset += FLASH_BLOCK_SIZE;
+
+      if (block_full && all_data_received) {
+        // last block of data couldn't fit the valid flag, write it in next
+        // block
+        memset(norcow_write_buffer, 0xFF, sizeof(norcow_write_buffer));
+        ((uint8_t *)norcow_write_buffer)[0] = NORCOW_VALID_FLAG;
+        ensure(flash_area_write_block(area, flash_offset, norcow_write_buffer),
+               NULL);
+        flash_offset += FLASH_BLOCK_SIZE;
+      }
+
+      norcow_write_buffer_filled = 0;
+      norcow_write_buffer_flashed += FLASH_BLOCK_SIZE;
+      memset(norcow_write_buffer, 0xFF, sizeof(norcow_write_buffer));
+
+      if (all_data_received) {
+        norcow_write_buffer_key = -1;
+        norcow_write_buffer_flashed = 0;
+      }
+    }
+  }
```
This looks OK, I think. Perhaps there is a cleaner way to write it, but I don't have time to think about it at the moment. We should test this code very thoroughly in the unit tests by creating and reading encrypted entries of all lengths between 0 to 132 bytes (so that up to two cipher blocks + another quadword are written).

TychoVrahe replied:
added the tests for the suggested lengths (for encrypted and unencrypted items too). I also fixed the python norcow_qr tests

https://github.com/trezor/trezor-firmware/commit/f5502791fdf64a5fa8a759cc6291738acb0294c8
https://github.com/trezor/trezor-firmware/commit/dfece8f91acdcac177bd901af7ff159a490292ea

onvej-sl replied:
@andrewkozlik Can we resolve this conversation?

andrewkozlik reviewed (changes requested):
I reviewed the norcow code (norcow.c, norcow-bitwise.h, norcow-blockwise.h). The build scripts also look correct. I haven't looked at the storage.c changes and I only glanced over the tests. @onvej-sl please take over the review, mainly storage.c and the tests, the rest of the code only if you feel it necessary.

onvej-sl commented:
What was the motivation behind changing the format of encrypted entries from `iv || tag || ciphertext` to `iv || ciphertext || tag`?

onvej-sl commented:
I'm not sure if I understand the blockwise flash. Is the following statement true? The blockwise flash does not support changing ones to zeros without resetting an entire block, which wears out the memory.

TychoVrahe commented:
@onvej-sl The previous format was incompatible with quadwords - the IV has 12 bytes, so it needed to be padded to 16B so that the tag can be written at later time. This is somewhat unpractical as it requires the storage to know and care about the underlying flash properties. When the tag is placed after the data, we can stream the bytes sequentially and handle the flash difference in norcow (buffering the bytes when incomplete quadwords are written). Like this, we can have unified handling of the encrypted data in storage.c. There is also detailed discussion about this format in above comments.

About the blockwise flash: almost correct, you can only modify data in the block (in case of STM32U5 this means quadword 4*32bit) when the block is erased (contains all 1) or if the new value is all zeroes. Erasing the block requires erasure of the whole flash page, which wears out the flash.

onvej-sl commented on storage/storage.c:
```diff
@@ -206,8 +204,8 @@ static secbool secequal(const void *ptr1, const void *ptr2, size_t n) {
   assert(n % sizeof(uint32_t) == 0);
-  assert((uintptr_t)ptr1 % sizeof(uint32_t) == 0);
-  assert((uintptr_t)ptr2 % sizeof(uint32_t) == 0);
+  //  assert((uintptr_t)ptr1 % sizeof(uint32_t) == 0);
+  //  assert((uintptr_t)ptr2 % sizeof(uint32_t) == 0);
```
Why the assertions were commented out?

TychoVrahe replied:
Restored https://github.com/trezor/trezor-firmware/commit/1afca92292fed9de3476bd0ba90ee1c0fc1600b2

(it was commented out because one of the previous iteration of the blockwise norcow had valid flag (one byte flag) at the beginning of the data, thus the data itself were not 32 bit aligned. Note that MCUs we currently use can handle unaligned access so the check is not strictly necessary)

onvej-sl commented on storage/storage.c:
```diff
@@ -390,34 +388,21 @@ static secbool set_wipe_code(const uint8_t *wipe_code, size_t wipe_code_len) {
   // The format of the WIPE_CODE_DATA_KEY entry is:
   // wipe code (variable), random salt (16 bytes), authentication tag (16 bytes)
```
The comment should read:  `  // wipe code (variable), random salt (8 bytes), authentication tag (8 bytes)`
The typo wasn't introduced in this pull request, but we can fix it anyway.

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/commit/1afca92292fed9de3476bd0ba90ee1c0fc1600b2

datis7756 replied:
- [ ] @

onvej-sl commented on storage/norcow.h:
```diff
@@ -65,18 +65,17 @@ secbool norcow_set_ex(uint16_t key, const void *val, uint16_t len,
- * Update the value of the given key starting at the given offset.
+ * Update the value of the given key, data are written sequentially from start
+ * Data are guaranteed to be stored on flash once the total item len is reached.
  * Note that you can only change bits from 1 to 0.
```
If I understand [this post](https://github.com/trezor/trezor-firmware/pull/3228#issuecomment-1912360454) correctly, the comment above is not true anymore.

TychoVrahe replied:
technically still true, though not sufficient. updated to better reflect the actual requirement

https://github.com/trezor/trezor-firmware/commit/1afca92292fed9de3476bd0ba90ee1c0fc1600b2

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,333 @@
+    val_offset -= NORCOW_LEN_LEN + NORCOW_KEY_LEN;
+  }
+
+  uint32_t end = val_offset + len + NORCOW_VALID_FLAG_LEN;
```
fixed https://github.com/trezor/trezor-firmware/pull/3228/commits/ed51043b140392879be04979fe27b6c7be7246a0

additionally similar bug fixed in https://github.com/trezor/trezor-firmware/commit/679e80ff184e4f7a3458bd4daba94b3832241951

onvej-sl replied:
@andrewkozlik Can you check it has been properly fixed?

TychoVrahe commented on storage/norcow_blockwise.h:
```diff
@@ -0,0 +1,333 @@
+        ensure(flash_area_write_block(&STORAGE_AREAS[sector], offset, block),
+               NULL);
+        offset += FLASH_BLOCK_SIZE;
+      }
```
changed https://github.com/trezor/trezor-firmware/pull/3228/commits/ed51043b140392879be04979fe27b6c7be7246a0

onvej-sl replied:
@andrewkozlik Can you check it has been properly fixed?

onvej-sl commented on storage/tests/c/flash.c:
```diff
@@ -146,3 +146,43 @@ secbool flash_write_word(uint16_t sector, uint32_t offset, uint32_t data) {
+
+secbool flash_write_block(uint16_t sector, uint32_t offset,
+                          const flash_block_t block) {
+#if FLASH_BLOCK_WORDS == 1 || defined FLASH_BIT_ACCESS
+  return flash_write_word(sector, offset, block[0]);
```
I don't understand why you assume that a block flash memory with one-word blocks automatically has bit access.

TychoVrahe replied:
I would say the assumption here is the other way around: flash with bit access has block size 1 word. This seems like an acceptable assumption to me, if you can write single bits you can also write single words, no need for blocks to be any larger.

onvej-sl replied:
> I would say the assumption here is the other way around: flash with bit access has block size 1 word.

This doesn't make sense to me. If `FLASH_BIT_ACCESS` is set, `FLASH_BLOCK_WORDS` is not used, which means that block size doesn't even need to be defined for a flash with bit access.

My complaint actually is that the code doesn't allow you to define a block flash with the block of size of one word. On the other hand, I don't think this is important, because don't have such a flash.

TychoVrahe replied:
ah i think i understand now

https://github.com/trezor/trezor-firmware/commit/540a4c2d133e52f9151e3f5a14624dfe29fb25c2

onvej-sl commented on storage/tests/python/src/norcow.py:
```diff
@@ -176,3 +165,166 @@ def _set_sectors(self, data):
+            raise ValueError("Norcow: no data on this offset")
+
+        key = self.sectors[self.active_sector][offset : offset + 2]
+        key = int.from_bytes(key, sys.byteorder)
```
The following issue wasn't introduced in this pull request but we may want to fix it.

In `_write`, little endian is used, while in `_read_item`, the endianness is retrieved from the system. This is inconsistent and causes that the tests won't work on big-endinan machines. We should either:
  * always use little-endian or
  * always use big-endian or
  * always retrieve the endianness from the system.
 
 By the way, we can use `key, length = struct.unpack('<HH', data)`.

TychoVrahe replied:
given that on the target MCUs we always have little endian and in tests we compare the memory bytewise, i went for _always use little-endian_ option. 

https://github.com/trezor/trezor-firmware/commit/106525346632493c313ccc068158ccc5bc0cb709

but this will likely still not work on big-endian machines, since the C code itself does not enforce little-endian and rely on underlying architecture. I don't think we should complicate the C code to make it enforce little endian, and if we don't do that, then failing the tests on big-endian machines might even be better, because when using sys.byteorder it would actually test data structure that is different then on target MCUs.

i also think that running these tests on big-endian machines is quite unlikely

onvej-sl replied:
> but this will likely still not work on big-endian machines, since the C code itself does not enforce little-endian and rely on underlying architecture

True, I didn't realize that.

It would be probably most consistent to replace all `sys.byteorder` with `'little'` and assert that `sys.byteorder == 'little'`, but it doesn't have to be done in this pull request.

onvej-sl commented on storage/tests/python/src/norcow.py:
```diff
@@ -176,3 +165,166 @@ def _set_sectors(self, data):
+class NorcowBlockwise(Norcow):
+    def __init__(self):
+        super().__init__()
+        self.word_size = 4 * consts.WORD_SIZE
```
Wouldn't be better to call it `block_size`?

onvej-sl commented on storage/tests/python/src/norcow.py:
```diff
@@ -176,3 +165,166 @@ def _set_sectors(self, data):
+
+    def _is_updatable(self, key: int, new: bytes) -> bool:
+        """
+        Item is updatable if the new value is the same or
+        it changes 1 to 0 only (the flash memory does not
+        allow to flip 0 to 1 unless you wipe it).
+
+        For flash with no bit access, the item is updatable
```
Unfinished comment, probably.

TychoVrahe replied:
and wrong too https://github.com/trezor/trezor-firmware/commit/57a01fb0a49289fd69b1e3fecb1677dc8fcdc91e

onvej-sl commented on storage/tests/python/src/norcow.py:
```diff
@@ -176,3 +165,166 @@ def _set_sectors(self, data):
+
+    def _write(self, pos: int, key: int, new_value: bytes) -> int:
+
+        if len(new_value) <= 12:
```
You may want to define `small_item_size = 12`.

TychoVrahe replied:
done https://github.com/trezor/trezor-firmware/commit/ecbcc1b7a07f560c645be4667405b88359853d83

onvej-sl commented on storage/tests/tests/test_set_get.py:
```diff
@@ -13,8 +14,57 @@
 
-def test_set_get():
-    sc, sp = common.init(unlock=True)
+@pytest.mark.parametrize("nc_class", NC_CLASSES)
+def test_set_delete(nc_class):
```
Just for information: You can create a fixture and parametrize the fixture instead of the test:
```python
@pytest.fixture(params=NC_CLASSES)
def nc_class(request):
    return request.param

def test_set_delete(nc_class):
```

Feel free to resolve this conversation yourself.

onvej-sl commented:
Do we have a test that tests the upgrade from version 3 to version 4?

TychoVrahe commented on storage/storage.c:
```diff
@@ -723,16 +712,23 @@ static secbool pin_fails_reset(void) {
+    if (sectrue != norcow_set(PIN_LOGS_KEY, new_logs, sizeof(new_logs))) {
+      return secfalse;
+    }
+  }
```
removed https://github.com/trezor/trezor-firmware/commit/1afca92292fed9de3476bd0ba90ee1c0fc1600b2

onvej-sl commented:
> Discussed with @andrewkozlik previously, we have one more modification to do, that is replace pin logs with some simpler, yet to be designed, form. The current form is very inefficient to work with on flash without bit access, and the protection itself is somewhat outdated given the vulnerabilities of STM32F4

We have 3 options:
  (a) Preserve the old format of pin logs in bitwise flash. Use the old format of pin logs in blockwise flash.
  (b) Design a new format of pin logs and use it in blockwise flash. Preserve the old format of pin logs in bitwise flash.
  (c) Design a new format of pin logs and use it in both bitwise and blockwise flash.


Andrew discussed this topic with me a bit but it seems to me that he concluded that it wasn't worth changing the pin logs. Do we want do design it without Andrew?

onvej-sl commented on storage/tests/tests/test_compact.py:
```diff
@@ -1,17 +1,19 @@
         s.set(0xBEEF, b"asdasdasdasd")
         s.set(0xBEEF, b"fsdasdasdasdasdsadasdsadasdasd")
-        s.set(0x0101, b"a" * (consts.NORCOW_SECTOR_SIZE - 600))
+        s.set(0x0101, b"a" * (consts.NORCOW_SECTOR_SIZE - 1200))
```
I assume the constant was changed because the test was failing with the old one. But why the test was failing with the old constant?

TychoVrahe replied:
on the blockwise (with 4word blocks) flash the items take more space so in order for all the requested item to fit we need larger reserve.

this however caused the bitwise test not to test compacting properly, so i added more parameters, and also asserts to ensure that the compacting is indeed tested: 

https://github.com/trezor/trezor-firmware/commit/ae40dd3108e31de93be3217404cdffebe7aa0534

onvej-sl commented:
> Discussed with @andrewkozlik previously, we have one more modification to do, that is replace pin logs with some simpler, yet to be designed, form. The current form is very inefficient to work with on flash without bit access, and the protection itself is somewhat outdated given the vulnerabilities of STM32F4

I discussed this topic a bit with Andrew, but I don't remember him sharing his plans about it with me.

We have these options:
  1. Leave it as it is. That is, use the current format of pin logs in both bitwise and blockwise flash. 

      Disadvantages:
        * The current format of pin logs would be inefficient to work with blockwise flash.

     Advantages:
        * Already implemented.
        * There is no need to implement update from the current format of pin logs to a new format.
        * There is no need to design and implement a new format.
  2. Use pair of ordinary counters in both bitwise and blockwise flash.

      Disadvantages:
        * Pin logs in both bitwise and blockwise flash wouldn't be resistant against fault injection attacks.
        * Update from the current format to a new format would have to be implemented.

      Advantages:
        * The implementation of the current format of pin logs can be removed.
        * Easy to implement.
  3. Preserve the current format of pin logs in bitwise flash. Use pair of ordinary counters in blockwise flash.

      Disadvantages:
        * Pin logs in blockwise flash wouldn't be resistant against fault injection attacks.

      Advantages:
        * There is no need to implement update from the current format of pin logs to a new format.
        * There is no need to design and implement a new format.
  4. Preserve the current format of pin logs in bitwise flash. Design a new format of pin logs and use it in blockwise flash.

      Disadvantages:
        * The new format of pin logs would have to be designed and implemented.

      Advantages:
        * There is no need to implement update from the current format of pin logs to a new format.

The question is whether the logs must be resistant to fault injection attacks. If they are not, an attacker could reset or decrement the pin counter and try more than 16 pins. Nevertheless, since the STM32F4 is vulnerable to flash read attacks, an attacker who performs this kind of attack can try as many pins as they want. Are we able to answer these questions?
  * Which of these two attacks is easier to execute?
  * Does it vary depending on the chip used?
  * Does it vary depending on the flash used?

onvej-sl commented:
I realized that the readme needs to be updated to reflect the fact that the order of the ciphertext and the tag has changed.

TychoVrahe commented:
updated the readme to better reflect current structure and separation between norcow and storage. The readme was kind of written to describe changes between some old format and never one, in sections i touched i tried to change this to describe current state.
https://github.com/trezor/trezor-firmware/pull/3228/commits/903f4d86f1c442a7c2eb10342d2f5baee3b25b62
https://github.com/trezor/trezor-firmware/pull/3228/commits/fa85d4ea17b0f84cc661a5ff775403a13f676fb6

TychoVrahe commented:
Regarding the pin logs:

ad 2): we still need to keep at least part of the current implementation of pin logs for upgrading purposes. 
(maybe we should introduce a mechanism to get rid of old code at least on models where the old storage format was never used)

- i dont know the mechanism of fault injection attacks to be able to say what it is easier, but given that flash read attacks on F2 and F4 are well known i would say we can assume they are easier or at least easy enough
- dependent on flash: fault injection attacks on STM32U5/blockwise flash should be harder due to the ECC protection, i guess
- dependent on chip: we expect glitching/flash read attacks on STM32U5 to be more difficult, so far this has not been tested

onvej-sl commented:
This idea occurred to me: If we wanted to go with option 4, we could implement the pin counter in blockwise memory in the same way as in bitwise memory, with the only difference being that we would set `PIN_LOG_WORDS` to `1`. One word is able record exactly 16 attempts. The pin log would have 12 = 4*(1+2*1) bytes, which would fit exactly into one block. It seems to me that it can't get any better in terms of memory, the loss of one block in case of an unsuccessful attempt and two blocks in a successful one seems inevitable to me. And we wouldn't need to design or implement anything new.

andrewkozlik commented:
Regarding the PIN logs, I think option 4 is the safest and most correct way to go about this. However, I am not strictly against the other options. What @onvej-sl  proposed with setting  the word count to 1 sounds interesting, although I can't fully judge it ATM.

IIRC, to protect against the sort of attacks that we were aiming against in the existing design, all we need in the PIN counter is some form of redundancy. The existing design was additionally aiming to be updatable by bit flips 1->0, which we don't need in the blockwise flash, and which is what makes the design so complex. I propose that we either (4a) remove the counter from blockwise storage and rely on the SE for the PIN counter in all future versions of Trezor, or (4b) implement something basic, like encode the fail count (0 to 15) into one byte, e.g. `count << 4 | ~count`, and repeat that byte as many times as it takes to fill a block. When the time comes to decommission the devices that use bitwise flash, then we can get rid of the complex PIN log code altogether, instead of being forced to carry it indefinitely.

onvej-sl commented:
I put together some my thoughts [here](https://www.notion.so/satoshilabs/PIN-log-64eee02021ac4e5d85cd1028c4a59342).

TychoVrahe commented:
In https://github.com/trezor/trezor-firmware/commit/a446550c458d3d95e1787bcd05c58ddea536ec67, i prototyped the "proposal 4" from notion. (i chose 4 as it seemed to me like reasonable balance of simplicity and protection, in case we select different implementation we can use the necessary structural changes from this commit)

Few notes:
- readme needs update again, will do after final algo selection
- crc16 impl is single purpose, maybe we should consider moving it to trezor-crypto along with existing crc32 implementation in extmod, or we can do this in separate PR
- existing upgrade test is limited to `NorcowBitwise`, makes no sense for blockwise.

onvej-sl commented on storage/pinlogs_blockwise.h:
```diff
@@ -0,0 +1,76 @@
+// The length of a word in bytes.
+#define WORD_SIZE (sizeof(uint32_t))
+
+uint16_t crc16(const uint8_t *data, uint32_t length) {
+  //  poly=0x1021, init=0xFFFF, xorout=0x0000 = CRC-16/CCITT-FALSE
+  uint8_t x;
+  uint16_t crc = 0xFFFF;
+
+  while (length--) {
+    x = crc >> 8 ^ *data++;
+    x ^= x >> 4;
+    crc = (crc << 8) ^ ((uint16_t)(x << 12)) ^ ((uint16_t)(x << 5)) ^
+          ((uint16_t)x);
+  }
+  return crc;
+}
```
I agree that it would be better to move it to trezor-crypto, but we don't have to do it in this pull request. I would just make it `static` for now.

By the way, since we have 4 bytes left in the block, we could use CRC32 instead of CRC16, if it had some benefit for us.

TychoVrahe replied:
made static here: https://github.com/trezor/trezor-firmware/pull/3228/commits/e280c472a3ed90e880bcb81bc017d2e3a18b14a2

we can change to crc32, or add some different redundancy (e.g. add crc32 along with the crc16, or copy the key and the counter or something?)

onvej-sl replied:
> we can change to crc32, or add some different redundancy (e.g. add crc32 along with the crc16, or copy the key and the counter or something?)

All options make sense, but I am unable to decide which one could give us the best benefit.

TychoVrahe replied:
added a repetition of the counter into the free bytes - should be better than leaving these empty, but i don't know whats best either

https://github.com/trezor/trezor-firmware/commit/2f1e91e1d738ef021f97b9eac6834db6eea2699c

 (still 16 byte per counter although encoding it like andrew proposed make sense too).

onvej-sl replied:
It looks alright.

TychoVrahe commented on storage/pinlogs_blockwise.h:
```diff
@@ -0,0 +1,76 @@
+
+  uint16_t logs[4] = {0};
+  logs[0] = key;
+  logs[1] = ~key;
```
changing it in https://github.com/trezor/trezor-firmware/commit/e280c472a3ed90e880bcb81bc017d2e3a18b14a2 
- assuming `random_uniform(0xFFFFFFFF)` was meant to be used in key generation
- the proposed key generation gives 0b01 for 0b1 (instead of outlined 0x10), assuming its ok
- the proposed check_key function returns secbool, but assumes sectrue==0xaaaaaaaa which seems suboptimal. I have changed it to `return ((key ^ (key << 1)) & sectrue);` but for some other sectrue values the check might not actually be sufficient like this. Maybe `return (((key ^ (key << 1)) & 0xAAAAAAAA) == 0xAAAAAAAA) ? sectrue : secfalse;` would be the way to go here?

onvej-sl replied:
> assuming random_uniform(0xFFFFFFFF) was meant to be used in key generation

Yes, good catch.

> the proposed key generation gives 0b01 for 0b1 (instead of outlined 0x10), assuming its ok

True.

> the proposed check_key function returns secbool, but assumes sectrue==0xaaaaaaaa which seems suboptimal. I have changed it to return ((key ^ (key << 1)) & sectrue); but for some other sectrue values the check might not actually be sufficient like this. Maybe return (((key ^ (key << 1)) & 0xAAAAAAAA) == 0xAAAAAAAA) ? sectrue : secfalse; would be the way to go here?

I understand. We can simplify by adding `static_assert(sectrue = 0xAAAAAAAA);` so that we are alerted if the constant changes. The option, `return (((key ^ (key << 1)) & 0xAAAAAAAA) == 0xAAAAAAAA) ? sectrue : secfalse;` is ok as well. Initially, I wanted to avoid using conditional branching because it seemed more prone to fault injection attacks. However, I now think this isn't a good reason as conditional branching generally cannot be avoided.

TychoVrahe replied:
using multiplication to avoid conditional: https://github.com/trezor/trezor-firmware/commit/2f1e91e1d738ef021f97b9eac6834db6eea2699c

onvej-sl replied:
It looks alright.

TychoVrahe commented:
I created issues for stuff mentioned and usolved above, hopefully didn't miss much: 
https://github.com/trezor/trezor-firmware/issues/3525
https://github.com/trezor/trezor-firmware/issues/3526
https://github.com/trezor/trezor-firmware/issues/3527

onvej-sl commented:
Do you mean [this](https://github.com/trezor/trezor-firmware/blob/2f1e91e1d738ef021f97b9eac6834db6eea2699c/storage/tests/tests/test_upgrade.py#L48) test? I don't think it tests [this](https://github.com/trezor/trezor-firmware/blob/2f1e91e1d738ef021f97b9eac6834db6eea2699c/storage/storage.c#L1359-L1401) branch.

TychoVrahe commented:
Ah missed that completely. Adding upgrade test from version 3 here: https://github.com/trezor/trezor-firmware/commit/225a401eb17707f6470d5d81351b2f8b5c7d392e

- upgrades from versions 1, 2 are not explicitly tested, but the upgrade code should be tested as part of other test, so this is likely not needed?
- freezed v3 storage from current main, not sure if there is a better way with less code duplication.
- i did NOT freeze the crypto source code used in v3, should these be also frozen? could some future updates in crypto lib break the upgrade?

andrewkozlik commented:
Following our discussion IRL, assuming we wanted this mapping from 4 bit values to bytes:
```
0x0 (0b0000) -> 0xaa (0b10101010)
0x1 (0b0001) -> 0xa9 (0b10101001)
0x2 (0b0010) -> 0xa6 (0b10100110)
0x3 (0b0011) -> 0xa5 (0b10100101)
0x4 (0b0100) -> 0x9a (0b10011010)
0x5 (0b0101) -> 0x99 (0b10011001)
0x6 (0b0110) -> 0x96 (0b10010110)
0x7 (0b0111) -> 0x95 (0b10010101)
0x8 (0b1000) -> 0x6a (0b01101010)
0x9 (0b1001) -> 0x69 (0b01101001)
0xa (0b1010) -> 0x66 (0b01100110)
0xb (0b1011) -> 0x65 (0b01100101)
0xc (0b1100) -> 0x5a (0b01011010)
0xd (0b1101) -> 0x59 (0b01011001)
0xe (0b1110) -> 0x56 (0b01010110)
0xf (0b1111) -> 0x55 (0b01010101)
```
The mapping can be implemented by the following transformation:
```
c = ((c << 2) | c) & 0x33;
c = ((c << 1) | c) & 0x55;
c = ((c << 1) | c) ^ 0xaa;
```
And the inverse map can be implemented by the following transformation:
```
c = c & 0x55;
c = ((c >> 1) | c) & 0x33;
c = ((c >> 2) | c) & 0x0f;
```

TychoVrahe commented:
@andrewkozlik there is a problem with 4 bit counter:
- we are writing the counter before doing the verification
- at 15 fails, the storage is not yet wiped, 16th fail is written and only after unsuccessful pin verification the storage is wiped (on successful verification pin counter is reset)

So we need the ability to write 16 fails, which is out of range for 4 bit counter.

andrewkozlik commented:
16 fails can be represented by erasing the counter entry. If the 16th attempt is successful, then the counter will be written to 0 afterwards. If no counter is found then the storage is invalid and should be wiped.

TychoVrahe commented:
changed the blockwise pin logs to the simple form here: https://github.com/trezor/trezor-firmware/commit/428c5e7da2d2e0878ea8ad4aede496f7782a15cb

one potential weak spot is in this check:

  ```
if (fails >= PIN_MAX_TRIES) {
    norcow_delete(PIN_LOGS_KEY);
    return sectrue;
  }
```

When this `if` is skipped due to glitching, the counter value will actually be decreased. 
I added an additional check after expanding the counter (see below), but it still feels a bit less robust.
```
  if (fails != compress_counter(ctr)) {
    handle_fault("PIN logs increase failed");
    return secfalse;
  } 
```

andrewkozlik commented on storage/pinlogs_blockwise.h:
```diff
@@ -0,0 +1,88 @@
+  return c;
+}
+
+static uint8_t compress_counter(uint8_t c) {
```
As I mentioned in a comment in the commit, there should be a format check:
```suggestion
static uint8_t compress_counter(uint8_t c) {
  if ((c ^ (c << 1)) & 0xAA != 0xAA) {
    handle_fault("ill-formed counter");
  }
```

TychoVrahe replied:
added https://github.com/trezor/trezor-firmware/commit/918f6bcdb092d030c7959194fb4107ac0bfde8fb

andrewkozlik commented:
FTR, if we wanted a mapping from 8 bits to 16 bits in the same manner as [above](https://github.com/trezor/trezor-firmware/pull/3228#issuecomment-1939237465), then it is this:
```
c = ((c << 4) | c) & 0x0f0f;
c = ((c << 2) | c) & 0x3333;
c = ((c << 1) | c) & 0x5555;
c = ((c << 1) | c) ^ 0xaaaa;
```
And the inverse:
```
c = c & 0x5555;
c = ((c >> 1) | c) & 0x3333;
c = ((c >> 2) | c) & 0x0f0f;
c = ((c >> 4) | c) & 0x00ff;
```
The validation function would use use `0xaaaa` instead of `0xaa`.

I can also come up with functions that map 5 bits to 15 bits (e.g. 0 -> 010 and 1 -> 101) with nearly identical code complexity.

TychoVrahe commented:
tbh, not having to deal with the special "deleted" case would seem preferable to me.

onvej-sl commented on storage/pinlogs_blockwise.h:
```diff
@@ -0,0 +1,91 @@
+  for (uint8_t i = 0; i < PIN_LOG_BYTES; i++) {
+    wait_random();
+    correct_bytes_cnt += compress_counter(((uint8_t *)logs)[i]) == c;
+    *ctr = c;
```
Is `*ctr = c;` intentionally inside the for cycle?

TychoVrahe replied:
yes i wanted this to be done repeatedly to make potential glitching harder. Though i am not sure this won't get optimized away, i guess nothing really prevents it, so it probably is useless measure

TychoVrahe commented:
switched to 8bit counter here: https://github.com/trezor/trezor-firmware/pull/3228/commits/aca987e0918b4465aa8de319cf2e4b249eb9ddc5

and added a readme entry here: https://github.com/trezor/trezor-firmware/pull/3228/commits/db484db072d691356f3a0d9d642cd111398353e1

onvej-sl commented on storage/pinlogs_blockwise.h:
```diff
@@ -0,0 +1,86 @@
+    *ctr = c;
+  }
+
+  success = sectrue * (correct_bytes_cnt == PIN_LOG_HALFWORDS);
+
+  return success;
```
I think the variable `success` is not need anymore:
```c
  return sectrue * (correct_bytes_cnt == PIN_LOG_HALFWORDS);
```

andrewkozlik replied:
I'd say
```c
  if (correct_bytes_cnt != PIN_LOG_HALFWORDS) {
    handle_fault(...);
    return secfalse;
  }
  return sectrue;
```

onvej-sl replied:
I agree that what Andrew suggests is more intuitive, but I think the idea behind the current solution was to avoid conditional branching if it's not too complicated to do so.

TychoVrahe replied:
yes that was the intention. 

removed the variable: https://github.com/trezor/trezor-firmware/pull/3228/commits/e3e6fb962eac98a4896a5649c25c4d26988517e5

andrewkozlik replied:
1. Unused variable left behind.
2. We should be calling `handle_fault()` if there is an inconsistency in the storage, which is the case when `correct_bytes_cnt != PIN_LOG_HALFWORDS`.

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/620aeb4271e7aac55156ea1bf30b886558cc57b7

onvej-sl commented on storage/storage.c:
```diff
@@ -1659,8 +1360,42 @@ static secbool storage_upgrade(void) {
     while (sectrue == norcow_get_next(&offset, &key, &val, &len)) {
-      if (sectrue != norcow_set(key, val, len)) {
-        return secfalse;
+      if (NORCOW_VERSION >= 4 && norcow_active_version < 4) {
```
The condition `NORCOW_VERSION >= 4` is neither needed now, nor in the future. Is that right?

TychoVrahe replied:
right, should be always true from now on: https://github.com/trezor/trezor-firmware/pull/3228/commits/1cc571c546d6208e76f6fff8e0e168352dabf177

onvej-sl reviewed (approved):
I reviewed `storage/storage.c`, `storage/pinlogs_{bitwise,blockwise}.h`, `README.md`, `storage/tests/**` and it looks good to me.

Hannsek commented:
There are some unwanted diffs:
https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/6174983951/artifacts/test_ui_report/master_diff.html
https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/6174984000/artifacts/master_diff/master_diff.html

TychoVrahe commented:
it needs rebase i guess

TychoVrahe commented:
squashed & rebased. fixed one more bug discovered by CI (https://github.com/trezor/trezor-firmware/pull/3228/commits/51cea6e7f1b306f44e0c4bfe66f214c9244d0ca2) 

@andrewkozlik if there is nothing else from your side please approve

Hannsek commented:
![image](https://github.com/trezor/trezor-firmware/assets/57008159/401263aa-e555-4ff6-b41b-b7d3f2d67861)
